models/iaf_wang_2002.cpp

```diff
@@ -0,0 +1,516 @@
+
+nest::iaf_wang_2002::iaf_wang_2002()
+  : ArchivingNode()
+  , P_()
+  , S_( P_ )
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+/* ---------------------------------------------------------------------------
+ * Copy constructor for node
+ * --------------------------------------------------------------------------- */
+
+nest::iaf_wang_2002::iaf_wang_2002( const iaf_wang_2002& n_ )
+  : ArchivingNode( n_ )
+  , P_( n_.P_ )
+  , S_( n_.S_ )
+  , B_( n_.B_, *this )
+{
+}
+
+/* ---------------------------------------------------------------------------
+ * Destructor for node
+ * --------------------------------------------------------------------------- */
+
+nest::iaf_wang_2002::~iaf_wang_2002()
+{
+  // GSL structs may not have been allocated, so we need to protect destruction
+
+  if ( B_.s_ )
+  {
+    gsl_odeiv_step_free( B_.s_ );
+  }
+
+  if ( B_.c_ )
+  {
+    gsl_odeiv_control_free( B_.c_ );
+  }
+
+  if ( B_.e_ )
+  {
+    gsl_odeiv_evolve_free( B_.e_ );
+  }
+}
+
+/* ---------------------------------------------------------------------------
+ * Node initialization functions
+ * --------------------------------------------------------------------------- */
+
+void
+nest::iaf_wang_2002::init_state_()
+{
+}
+
+void
+nest::iaf_wang_2002::init_buffers_()
+{
+  B_.spikes_.resize( 3 );
+  for ( auto& sb : B_.spikes_ )
+  {
+    sb.clear(); // includes resize
+  }
+
+  B_.currents_.clear(); // includes resize
+
+  B_.logger_.reset(); // includes resize
+  ArchivingNode::clear_history();
+
+  if ( B_.s_ == nullptr )
+  {
+    B_.s_ = gsl_odeiv_step_alloc( gsl_odeiv_step_rkf45, State_::STATE_VEC_SIZE );
+  }
+  else
+  {
+    gsl_odeiv_step_reset( B_.s_ );
+  }
+
+  if ( B_.c_ == nullptr )
+  {
+    B_.c_ = gsl_odeiv_control_y_new( P_.gsl_error_tol, 0.0 );
+  }
+  else
+  {
+    gsl_odeiv_control_init( B_.c_, P_.gsl_error_tol, 0.0, 1.0, 0.0 );
+  }
+
+  if ( B_.e_ == nullptr )
+  {
+    B_.e_ = gsl_odeiv_evolve_alloc( State_::STATE_VEC_SIZE );
+  }
+  else
+  {
+    gsl_odeiv_evolve_reset( B_.e_ );
+  }
+
+  B_.sys_.function = iaf_wang_2002_dynamics;
+  B_.sys_.jacobian = nullptr;
+  B_.sys_.dimension = State_::STATE_VEC_SIZE;
+  B_.sys_.params = reinterpret_cast< void* >( this );
+  B_.step_ = Time::get_resolution().get_ms();
+  B_.integration_step_ = Time::get_resolution().get_ms();
+
+  B_.I_stim_ = 0.0;
+}
+
+void
+nest::iaf_wang_2002::pre_run_hook()
+{
+  // ensures initialization in case mm connected after Simulate
+  B_.logger_.init();
+
+  V_.RefractoryCounts_ = Time( Time::ms( P_.t_ref ) ).get_steps();
+  // since t_ref_ >= 0, this can only fail in error
+  assert( V_.RefractoryCounts_ >= 0 );
+
+  // helper vars
+  const double alpha_tau = P_.alpha * P_.tau_rise_NMDA;
+  const double tau_rise_tau_dec = P_.tau_rise_NMDA / P_.tau_decay_NMDA;
+
+  V_.S_jump_1 = exp( -P_.alpha * P_.tau_rise_NMDA ) - 1;
+  V_.S_jump_0 = pow( alpha_tau, tau_rise_tau_dec ) * boost::math::tgamma_lower( 1 - tau_rise_tau_dec, alpha_tau );
```

Comment:
NEST guidelines suggest to always use `std::` and `expm1()` is more precise.

```suggestion
  V_.S_jump_1 = std::expm1( -P_.alpha * P_.tau_rise_NMDA );
  V_.S_jump_0 = std::pow( alpha_tau, tau_rise_tau_dec ) * boost::math::tgamma_lower( 1 - tau_rise_tau_dec, alpha_tau );
```